PWGHF/D2H/Tasks/taskB0.cxx

```diff
@@ -159,28 +105,71 @@ struct HfTaskB0Mc {
+  {
+    for (auto const& candidate : candidates) {
+      if (!TESTBIT(candidate.hfflag(), hf_cand_b0::DecayType::B0ToDPi)) {
+        continue;
+      }
+      if (yCandMax >= 0. && std::abs(yB0(candidate)) > yCandMax) {
+        continue;
+      }
+
+      auto candD = candidate.prong0_as<soa::Join<aod::HfCand3Prong, aod::HfSelDplusToPiKPi>>();
+      // auto candPi = candidate.prong1_as<TracksWithSel>();
+
+      registry.fill(HIST("hMass"), invMassB0ToDPi(candidate), candidate.pt());
+      registry.fill(HIST("hPtCand"), candidate.pt());
+      registry.fill(HIST("hPtProng0"), candidate.ptProng0());
+      registry.fill(HIST("hPtProng1"), candidate.ptProng1());
+      registry.fill(HIST("hIPProd"), candidate.impactParameterProduct(), candidate.pt());
+      registry.fill(HIST("hDecLength"), candidate.decayLength(), candidate.pt());
+      registry.fill(HIST("hDecLengthXY"), candidate.decayLengthXY(), candidate.pt());
+      registry.fill(HIST("hd0Prong0"), candidate.impactParameter0(), candidate.pt());
+      registry.fill(HIST("hd0Prong1"), candidate.impactParameter1(), candidate.pt());
+      registry.fill(HIST("hCPA"), candidate.cpa(), candidate.pt());
+      registry.fill(HIST("hEta"), candidate.eta(), candidate.pt());
+      registry.fill(HIST("hRapidity"), yB0(candidate), candidate.pt());
+      registry.fill(HIST("hImpParErr"), candidate.errorImpactParameter0(), candidate.pt());
+      registry.fill(HIST("hImpParErr"), candidate.errorImpactParameter1(), candidate.pt());
+      registry.fill(HIST("hDecLenErr"), candidate.errorDecayLength(), candidate.pt());
+      registry.fill(HIST("hDecLenXYErr"), candidate.errorDecayLengthXY(), candidate.pt());
+      registry.fill(HIST("hInvMassD"), invMassDplusToPiKPi(candD), candidate.pt());
+    } // candidate loop
+  }   // process
+
+  /// B0 MC analysis and fill histograms
   void processMc(soa::Filtered<soa::Join<aod::HfCandB0, aod::HfSelB0ToDPi, aod::HfCandB0McRec>> const& candidates,
-                 soa::Join<aod::McParticles, aod::HfCandB0McGen> const& particlesMC, aod::BigTracksMC const& tracks, aod::HfCand3Prong const&)
+                 soa::Join<aod::McParticles, aod::HfCandB0McGen> const& particlesMC,
+                 aod::BigTracksMC const& tracks,
+                 soa::Join<aod::HfCand3Prong, aod::HfCand3ProngMcRec> const&)
   {
     // MC rec
     for (auto const& candidate : candidates) {
-      if (TESTBIT(candidate.hfflag(), hf_cand_b0::DecayType::B0ToDPi)) {
+      if (!TESTBIT(candidate.hfflag(), hf_cand_b0::DecayType::B0ToDPi)) {
         continue;
       }
       if (yCandMax >= 0. && std::abs(yB0(candidate)) > yCandMax) {
         continue;
       }
-      auto candD = candidate.prong0_as<aod::HfCand3Prong>();
+      // auto candD = candidate.prong0_as<aod::HfCand3Prong>();
+      auto candD = candidate.prong0_as<soa::Join<aod::HfCand3Prong, aod::HfCand3ProngMcRec>>();
       if (TESTBIT(std::abs(candidate.flagMcMatchRec()), hf_cand_b0::DecayType::B0ToDPi)) {
```

Comment:
Are you expecting to use `candD` also for unmatched candidates? If not, I suggest to move it inside the `TESTBIT` block, since otherwise it is executed unnecessarily taking some extra CPU spent on the casting, AFAIK.
In any case, it is an unused variable now and will not pass the compilation checks.

Reply:
`candD` is called in both `if` (signal) and `else` (background) blocks, that's why it is defined above.